### ISSUE TITLE
Some emojis are replaced with a substituted character

### ISSUE THREAD

**Issue by brandondrew** [zed-industries/zed#16002]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Although there are a few other bug reports related to emojis, none appear to describe the behavior I'm experiencing. 

Most emojis display just as expected in Zed, but every once in a while I run into one that fails to appear correctly.  In order to isolate the problem--whether it occurs on insertion or display--I created a file in BBEdit, which does not have the same problem as Zed.  I inserted 2 emojis in BBEdit and then opened it in Zed, and found Zed displayed it exactly as it does when inserting the same emoji into Zed, so the problem is related to display of that emoji and not insertion.  I added a single lower-case 'x' in Zed and saved the file to see if Zed would corrupt it by changing the emoji character it could not display correctly.  It did not.  The file appeared in BBEdit the same except that the lowercase x was added.

Steps to reproduce the problem:

1. create a file in BBEdit and add the red ❌ emoji (as well as others if you want);
2. open the file in Zed and notice that the emoji is replaced with a different character;
3. if you want to prove that Zed is only having problems with display, add characters and open the file again in BBEdit.

<img width="451" alt="image" src="https://github.com/user-attachments/assets/3f59a772-6f3b-4dea-9316-f8884288eb4f">

### Environment

Zed: v0.147.2 (Zed)
OS: macOS 14.5.0
Memory: 16 GiB
Architecture: aarch64

**Comment by JunkuiZhang** [comment]
Can not reproduce on my Mac, I copied and pasted ❌ from your description:

<img width="828" alt="截屏2024-08-12 00 49 09" src="https://github.com/user-attachments/assets/96907d45-f39b-4233-8f13-c7c83b913bd1">

**Comment by JunkuiZhang** [comment]
Would you mind sharing your `emoji.txt` file?

**Comment by brandondrew** [comment]
Here it is:

[emoji.txt](https://github.com/user-attachments/files/16576561/emoji.txt)

**Comment by brandondrew** [comment]
> Can not reproduce on my Mac

I'm on macOS 14.5 (23F79)... what version are you on?

**Comment by JunkuiZhang** [comment]
> I'm on macOS 14.5 (23F79)... what version are you on?

Same OS version, but I'm on the latest main branch `zed` build. I still can not reproduce this issue:

<img width="828" alt="截屏2024-08-12 18 46 02" src="https://github.com/user-attachments/assets/2858c410-e8a7-4c42-8cce-b3ccdf0ba5ce">

**Comment by dcontiveros** [comment]
I just reproduced this. The rendering depends on the context. Most text editors handle a majority of the subset of emojis. However, Zed does choke on crossmark. Reference the attached screenshot. I have also included a test file.

![ss](https://github.com/user-attachments/assets/24044d19-2f5d-4fdb-8e88-f1707a6ac005)


[test.txt](https://github.com/user-attachments/files/16621613/test.txt)

**Comment by JunkuiZhang** [comment]
Still can't reproduce on my end, this is so strange.

<img width="828" alt="截屏2024-08-15 13 31 10" src="https://github.com/user-attachments/assets/a1b35a0b-326e-487c-bcf1-4ecd8664933d">

**Comment by dcontiveros** [comment]
@JunkuiZhang what is the font in your `settings.json`?

**Comment by JunkuiZhang** [comment]
> @JunkuiZhang what is the font in your `settings.json`?

Arial

https://github.com/user-attachments/assets/343d0ebe-f5bc-458e-a9cc-9e6d43d6a1e3

**Comment by dcontiveros** [comment]
I've reproduced the problem. It appears to be an issue when using Zed Mono as the font.


https://github.com/user-attachments/assets/68e8eff1-5d12-4609-a5de-63c03760ef68

Using Andale Mono results in the proper rendering.

**Comment by brandondrew** [comment]
> > @JunkuiZhang what is the font in your `settings.json`?
> 
> Arial

I never ever ever ever use Arial <h1>😂</h1>

**Comment by brandondrew** [comment]
For me even Andale Mono fails.  (I was previously using the default font, which I think changed from Zed Mono to some IBM font recently.)

<img width="400" alt="image" src="https://github.com/user-attachments/assets/c1f5fa69-3569-4815-a353-5eda954dd374">

**Comment by brandondrew** [comment]
Arial fails too 😰

**Comment by dcontiveros** [comment]
> For me even Andale Mono fails.  (I was previously using the default font, which I think changed from Zed Mono to some IBM font recently.)
> 
> 
> 
> <img width="400" alt="image" src="https://github.com/user-attachments/assets/c1f5fa69-3569-4815-a353-5eda954dd374">
> 
> 


This is a different option. Try buffer font option.

**Comment by brandondrew** [comment]
Like this?
<img width="402" alt="image" src="https://github.com/user-attachments/assets/9e464ccb-97a5-48ca-a0d3-e19e7920a39a">

**Comment by brandondrew** [comment]
Got it!

<img width="431" alt="image" src="https://github.com/user-attachments/assets/e3e8dd01-d3de-4dd7-9fc7-556e0c0815fa">

**Comment by dcontiveros** [comment]
I believe this needs to be classified as an issue with Zed Mono. This isn't the only font that is afflicted. I am partial to Iosevka and have seen this issue as well.

**Comment by brandondrew** [comment]
@dcontiveros If it afflicts multiple fonts, then are you sure it is a flaw in all of those fonts?

**Comment by JunkuiZhang** [comment]
> @dcontiveros If it afflicts multiple fonts, then are you sure it is a flaw in all of those fonts?

The codepoint `274C`, which represents the character `❌`, is handled by `Zed Plex Mono`:

<img width="652" alt="Screenshot 2024-08-16 13 01 24" src="https://github.com/user-attachments/assets/db1c4e39-4827-4867-a9df-456af85fba83">

Therefore, the glyph display issue is a problem with the font, not with Zed. If a font like `Arial`, which does not include this codepoint, is used, Zed will fallback to `CoreText` for rendering the character correctly.

**Comment by dcontiveros** [comment]
Yes I figured this would most likely be a font issue vs. drawing. Emoji's are tough, especially in terminals. Any chance we can get a change on Zed Mono ?

**Comment by JunkuiZhang** [comment]
> Yes I figured this would most likely be a font issue vs. drawing. Emoji's are tough, especially in terminals. Any chance we can get a change on Zed Mono ?

Yes, simply deleting that codepoint can work, here is a quick demo, I deleted that glyph from `Zed Plex Mono`:


<img width="828" alt="截屏2024-08-16 14 15 24" src="https://github.com/user-attachments/assets/0535ec20-3f7e-4a00-9372-ed0f46c24457">

cc @notpeter Since there are some issue with `Zed Plex Mono` already in #15542, so I guess maybe could open an issue for the font?

**Comment by notpeter** [comment]
Yeah, I don't know if the process for creating Zed Plex is documented/automated. @danilo-leal do you know?

**Comment by dragnev-dev** [comment]
Sad to see the PR not getting merged.

Workaround: download the fonts from the PR, install them locally
